src/advanced/timestamping-tutorial.md

```diff
@@ -0,0 +1,929 @@
+  calculates the resulting hash of the blockchain state.
+
+As we will need to add entries to the database to record timestamps, we also
+declare two methods for a `Fork`-based schema:
+
+```rust
+impl<'a> Schema<&'a mut Fork> {
+    /// Returns the mutable `ProofMapIndex` of timestamps.
+    pub fn timestamps_mut(&mut self) -> ProofMapIndex<&mut Fork, Hash, TimestampEntry> {
+        ProofMapIndex::new("timestamping.timestamps", &mut self.view)
+    }
+
+    /// Adds the timestamp entry to the database.
+    pub fn add_timestamp(&mut self, timestamp_entry: TimestampEntry) {
+        let timestamp = timestamp_entry.timestamp();
+        let content_hash = timestamp.content_hash();
+
+        // Check that timestamp with given content_hash does not exist.
+        if self.timestamps().contains(content_hash) {
+            return;
+        }
+
+        // Add timestamp
+        self.timestamps_mut().put(content_hash, timestamp_entry);
+    }
+}
+```
+
+- the `timestamps_mut` method returns a mutable `ProofMapIndex` of the
+  timestamping map
+- the `add_timestamp` method  adds a new timestamp which contains the timestamp
+  entry which we have defined previously and the hash of the timestamp. This
+  method also checks if a timestamp with the given hash already
+  exists in the blockchain.
+
+## Configure Transactions
+
+[Transactions](../architecture/transactions.md) resemble messages and
+perform atomic actions on the blockchain state. We have separated the code that
+configures transactions for our Timestamping service into a separate file -
+[src/transactions.rs][src/transactions.rs]. The Timestamping service we are
+creating requires a single transaction type - adding timestamping transactions
+to the blockchain.
+
+First, we define the structures we need from Exonum core and the time oracle.
+Here we also indicate a connection to the schema we have configured previously
+in [src/schema.rs][src/schema.rs] and the `TIMESTAMPING_SERVICE` constant
+declared in [src/lib.rs][src/lib.rs].
+
+```rust
+use exonum::{
+    blockchain::{ExecutionError, ExecutionResult, Transaction}, crypto::{CryptoHash, PublicKey},
+    messages::Message, storage::Fork,
+};
+use exonum_time::schema::TimeSchema;
+
+use schema::{Schema, Timestamp, TimestampEntry};
+use TIMESTAMPING_SERVICE;
+```
+
+Service transactions are defined through the `transactions!` macro which
+automatically assigns transaction IDs based on the declaration order:
+
+```rust
+transactions! {
+    pub TimeTransactions {
+        const SERVICE_ID = TIMESTAMPING_SERVICE;
+
+        /// A timestamp transaction.
+        struct TxTimestamp {
+            /// Public key of transaction.
+            pub_key: &PublicKey,
+
+            /// Timestamp content.
+            content: Timestamp,
+        }
+    }
+}
+```
+The transaction to create a timestamping transaction (`TxTimestamp`) contains
+the public key of the node, which is used for transaction encryption, and the
```

Comment:
This is not the public key of the node, and even not the public key of transaction, as stated in doc comment. It's the public key of the user.